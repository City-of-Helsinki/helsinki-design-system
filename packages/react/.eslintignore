**/*.d.ts
webpack.config.js
.templates
lib/
**/*.stories.tsx
