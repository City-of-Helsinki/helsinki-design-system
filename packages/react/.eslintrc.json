{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:jest/recommended",
    "airbnb",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended" // this should always be the last item on this array
  ],
  "plugins": ["react", "@typescript-eslint", "prettier", "jest", "import"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "ignorePatterns": ["*.module.css"],
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*.test.ts", "**/*.test.tsx", "**/test*.ts*", "**/*.stories.tsx"]
      }
    ],
    "import/order": [
      "error",
      {
        "groups": ["builtin", "external", "internal", ["parent", "sibling", "index"]],
        "newlines-between": "always"
      }
    ],
    "react/jsx-filename-extension": "off",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-curly-newline": "off",
    "@typescript-eslint/explicit-function-return-type": "off",
    "react/prop-types": "off",
    "import/extensions": "off",
    "import/prefer-default-export": "off",
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "components": ["Label"],
        "required": {
          "some": ["nesting", "id"]
        },
        "allowChildren": false
      }
    ],
    "@typescript-eslint/explicit-member-accessibility": "off",
    "react/state-in-constructor": [2, "never"],
    "no-unused-expressions": [2, {
      "allowTernary": true
    }],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": ["error"]
  },
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"],
        "paths": ["src"]
      }
    }
  },
  "env": {
    "browser": true,
    "jest": true
  },
  "globals": {
    "JSX": true
  }
}
