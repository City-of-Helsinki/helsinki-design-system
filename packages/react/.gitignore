node_modules/
storybook-static/
lib/
_bundles/
*.css.d.ts
.loki/current
.loki/difference

*storybook.log
