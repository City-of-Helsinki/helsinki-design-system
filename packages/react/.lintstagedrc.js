module.exports = {
  'src/**/*.{ts,tsx}': 'eslint',
}
