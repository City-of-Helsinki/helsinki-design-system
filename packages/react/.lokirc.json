{
  "chromeRetries": 2,
  "configurations": {
    "chrome.laptop": {
      "target": "chrome.docker",
      "width": 1366,
      "height": 768,
      "mobile": false
    },
    "chrome.iphone7": {
      "target": "chrome.docker",
      "preset": "iPhone 7"
    }
  },
  "skipStories": "Icons/.*|Button/*|Tag/*|Hero/*|Koros/*|Checkbox/*",
  "pixelmatch": {
    "threshold": 0.2,
    "alpha": 1,
    "diffColor": [255, 0, 255]
  }
}
