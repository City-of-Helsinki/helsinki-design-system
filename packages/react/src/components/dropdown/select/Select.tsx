/* eslint-disable react/destructuring-assignment */
import React, { FocusEvent, useCallback, useRef, useState } from 'react';
import {
  useSelect,
  useMultipleSelection,
  A11yRemovalMessage,
  A11yStatusMessageOptions,
  UseMultipleSelectionStateChange,
  UseMultipleSelectionStateChangeOptions,
  UseMultipleSelectionState,
} from 'downshift';
import isEqual from 'lodash.isequal';
import uniqueId from 'lodash.uniqueid';
import { useVirtual } from 'react-virtual';

import 'hds-core';

import styles from './Select.module.scss';
import { FieldLabel } from '../../../internal/field-label/FieldLabel';
import classNames from '../../../utils/classNames';
import { IconAlertCircle, IconAngleDown, IconCrossCircle } from '../../../icons';
import { SelectedItems } from '../../../internal/selectedItems/SelectedItems';
import { DROPDOWN_MENU_ITEM_HEIGHT, getIsInSelectedOptions } from '../dropdownUtils';
import { DropdownMenu } from '../../../internal/dropdownMenu/DropdownMenu';
import getIsElementFocused from '../../../utils/getIsElementFocused';
import getIsElementBlurred from '../../../utils/getIsElementBlurred';
import { useTheme } from '../../../hooks/useTheme';

export interface SelectCustomTheme {
  '--dropdown-background-default'?: string;
  '--dropdown-background-disabled'?: string;
  '--dropdown-border-color-default'?: string;
  '--dropdown-border-color-hover'?: string;
  '--dropdown-border-color-hover-invalid'?: string;
  '--dropdown-border-color-focus'?: string;
  '--dropdown-border-color-invalid'?: string;
  '--dropdown-border-color-disabled'?: string;
  '--dropdown-color-default'?: string;
  '--dropdown-color-disabled'?: string;
  '--focus-outline-color'?: string;
  '--helper-color-default'?: string;
  '--helper-color-invalid'?: string;
  '--menu-divider-color'?: string;
  '--menu-item-background-default'?: string;
  '--menu-item-background-hover'?: string;
  '--menu-item-background-selected'?: string;
  '--menu-item-background-selected-hover'?: string;
  '--menu-item-background-disabled'?: string;
  '--menu-item-color-default'?: string;
  '--menu-item-color-hover'?: string;
  '--menu-item-color-selected'?: string;
  '--menu-item-color-selected-hover'?: string;
  '--menu-item-color-disabled'?: string;
  '--menu-item-icon-color-selected'?: string;
  '--menu-item-icon-color-disabled'?: string;
  '--multiselect-checkbox-background-selected'?: string;
  '--multiselect-checkbox-background-disabled'?: string;
  '--multiselect-checkbox-border-default'?: string;
  '--multiselect-checkbox-border-hover'?: string;
  '--multiselect-checkbox-border-disabled'?: string;
  '--multiselect-checkbox-color-default'?: string;
  '--multiselect-checkbox-color-selected'?: string;
  '--multiselect-checkbox-color-selected-disabled'?: string;
  '--placeholder-color'?: string;
}

export type CommonSelectProps<OptionType> = {
  /**
   * When `true`, allows moving from the first item to the last item with Arrow Up, and vice versa using Arrow Down.
   */
  circularNavigation?: boolean;
  /**
   * Additional class names to apply to the select
   */
  className?: string;
  /**
   * Flag for whether the clear selections button should be displayed
   */
  clearable?: boolean;
  /**
   * If `true`, the dropdown will be disabled
   */
  disabled?: boolean;
  /**
   * Function used to generate an ARIA a11y message when an item is selected. See [here](https://github.com/downshift-js/downshift/tree/master/src/hooks/useSelect#geta11yselectionmessage) for more information.
   */
  getA11ySelectionMessage?: (options: A11yStatusMessageOptions<OptionType>) => string;
  /**
   * Function used to generate an ARIA a11y message when the status changes. See [here](https://github.com/downshift-js/downshift/tree/master/src/hooks/useSelect#geta11ystatusmessage) for more information.
   */
  getA11yStatusMessage?: (options: A11yStatusMessageOptions<OptionType>) => string;
  /**
   * A helper text that will be shown below the dropdown
   */
  helper?: React.ReactNode;
  /**
   * An error text that will be shown below the dropdown when `invalid` is true
   */
  error?: React.ReactNode;
  /**
   * Icon to be shown in the dropdown
   */
  icon?: React.ReactNode;
  /**
   * Used to generate the first part of the id on the elements
   */
  id?: string;
  /**
   * If `true`, the input and `helper` will be displayed in an invalid state
   */
  invalid?: boolean;
  /**
   * A function used to detect whether an option is disabled
   */
  isOptionDisabled?: (option: OptionType, index: number) => boolean;
  /**
   * The label for the dropdown
   */
  label: React.ReactNode;
  /**
   * Callback function fired when the state is changed
   */
  onBlur?: () => void;
  /**
   * Callback function fired when the component is focused
   */
  onFocus?: () => void;
  /**
   * Sets the data item field that represents the item label
   * E.g. an `optionLabelField` value of `'foo'` and a data item `{ foo: 'Label', bar: 'value' }`, would display `Label` in the menu for that specific item
   */
  optionLabelField?: string;
  /**
   * Array of options that should be shown in the menu
   */
  options: OptionType[];
  /**
   * Short hint displayed in the dropdown before the user enters a value
   */
  placeholder?: string;
  /**
   * If `true`, marks the dropdown as required
   */
  required?: boolean;
  /**
   * Override or extend the root styles applied to the component
   */
  style?: React.CSSProperties;
  /**
   * Custom theme styles
   */
  theme?: SelectCustomTheme;
  /**
   * If `true`, the menu options will be virtualized. This greatly increases performance when there are a lot of options,
   * but screen readers won't be able to know how many options there are.
   */
  virtualized?: boolean;
  /**
   * Sets the number of options that are visible in the menu before it becomes scrollable
   */
  visibleOptions?: number;
  /**
   * Aria-label text for the tooltip
   */
  tooltipLabel?: string;
  /**
   * Aria-label text for the tooltip trigger button
   */
  tooltipButtonLabel?: string;
  /**
   * The text content of the tooltip
   */
  tooltipText?: string;
};

export type SingleSelectProps<OptionType> = CommonSelectProps<OptionType> & {
  /**
   * The aria-label for the clear button
   */
  clearButtonAriaLabel?: string;
  /**
   * When `true`, enables selecting multiple values
   */
  multiselect?: false;
  /**
   * Value that should be selected when the dropdown is initialized
   */
  defaultValue?: OptionType | null;
  /**
   * Callback function fired when the state is changed
   */
  onChange?: (selected: OptionType) => void;
  /**
   * The selected value
   */
  value?: OptionType | null;
};

export type MultiSelectProps<OptionType> = CommonSelectProps<OptionType> & {
  /**
   * The aria-label for the clear button
   */
  clearButtonAriaLabel: string;
  /**
   * When `true`, enables selecting multiple values
   */
  multiselect: true;
  /**
   * Value(s) that should be selected when the dropdown is initialized
   */
  defaultValue?: OptionType[] | null;
  /**
   * Function used to generate an ARIA a11y message when an item is removed. See [here](https://github.com/downshift-js/downshift/tree/master/src/hooks/useMultipleSelection#geta11yremovalmessage) for more information.
   */
  getA11yRemovalMessage?: (options: A11yRemovalMessage<OptionType>) => string;
  /**
   * Callback function fired when the state is changed
   */
  onChange?: (selected: OptionType[]) => void;
  /**
   * The aria-label for the selected item remove button.
   * You can use a special {value} token that will be replaced with the actual item value.
   * E.g. an item with the label Foo and property value of `'Remove ${value}'` would become `aria-label="Remove Foo"`.
   */
  selectedItemRemoveButtonAriaLabel: string;
  /**
   * A label for the selected items that is only visible to screen readers. Can be used to to give screen reader users additional information about the selected item.
   * You can use a special {value} token that will be replaced with the actual item value.
   * E.g. an item with the label Foo and property value of `'Selected item ${value}'` would become `aria-label="Selected item Foo"`.
   */
  selectedItemSrLabel?: string;
  /**
   * The selected value(s)
   */
  value?: OptionType[] | null;
};

export type SelectProps<OptionType> = SingleSelectProps<OptionType> | MultiSelectProps<OptionType>;

/**
 * Multi-select state change handler
 * @param type
 * @param activeIndex
 * @param currentActiveIndex
 * @param selectedItemsContainerEl
 */
export function onMultiSelectStateChange<T>(
  { type, activeIndex }: UseMultipleSelectionStateChange<T>,
  currentActiveIndex,
  selectedItemsContainerEl,
) {
  let activeNode;
  const { FunctionRemoveSelectedItem, SelectedItemKeyDownBackspace } = useMultipleSelection.stateChangeTypes;

  if (type === FunctionRemoveSelectedItem || type === SelectedItemKeyDownBackspace) {
    activeNode = selectedItemsContainerEl?.childNodes[currentActiveIndex] as HTMLDivElement;
    if (!activeIndex && activeNode) activeNode.focus();
  }
}

/**
 * Multi-select reducer function
 * @param state
 * @param type
 * @param changes
 * @param controlled
 */
export function multiSelectReducer<T>(
  state: UseMultipleSelectionState<T>,
  { type, changes }: UseMultipleSelectionStateChangeOptions<T>,
  controlled: boolean,
) {
  const { FunctionRemoveSelectedItem, SelectedItemKeyDownBackspace } = useMultipleSelection.stateChangeTypes;

  if (type === FunctionRemoveSelectedItem || type === SelectedItemKeyDownBackspace) {
    // get the index of the deleted item
    const removedItemIndex = state.selectedItems.findIndex((item) => !changes.selectedItems.includes(item));
    // activeIndex updates at a different time depending on whether the
    // component is controlled. For controlled components, when 'onStateChange'
    // is called, the removed item still has a SelectedItem representing
    // it in the DOM. This means that the succeeding SelectedItem is at
    // index n + 1 instead of n
    const adjustedIndex = controlled ? removedItemIndex + 1 : removedItemIndex;
    // whether the removed item was last in selectedItems
    const lastItemRemoved = removedItemIndex === changes.selectedItems.length;

    return {
      ...changes,
      // set the new last item as active if the removed item was last,
      // otherwise set the item succeeding the removed one active
      activeIndex: lastItemRemoved ? removedItemIndex - 1 : adjustedIndex,
    };
  }

  return changes;
}

export const Select = <OptionType,>(props: SelectProps<OptionType>) => {
  // we can't destructure all the props. after destructuring, the link
  // between the multiselect prop and the value, onChange etc. props would vanish
  const {
    circularNavigation = false,
    className,
    clearable = props.multiselect,
    disabled = false,
    error,
    getA11ySelectionMessage = () => '',
    getA11yStatusMessage = () => '',
    helper,
    id = uniqueId('hds-select-') as string,
    invalid,
    isOptionDisabled,
    label,
    onBlur = () => null,
    onFocus = () => null,
    optionLabelField = 'label',
    options = [],
    placeholder,
    required,
    style,
    theme,
    virtualized = false,
    visibleOptions = 5,
    tooltipLabel,
    tooltipButtonLabel,
    tooltipText,
  } = props;
  // flag for whether the component is controlled
  const controlled = props.multiselect && props.value !== undefined;
  // custom theme class that is applied to the root element
  const customThemeClass = useTheme<SelectCustomTheme>(styles.root, theme);
  // selected items container ref
  const selectedItemsContainerRef = useRef<HTMLDivElement>();
  // menu ref
  const menuRef = React.useRef<HTMLUListElement>();
  // toggle button ref
  const toggleButtonRef = React.useRef<HTMLButtonElement>(null);
  // whether active focus is within the dropdown
  const [hasFocus, setFocus] = useState<boolean>(false);
  // virtualize menu items to increase performance
  const virtualizer = useVirtual<HTMLUListElement>({
    size: options.length,
    parentRef: menuRef,
    estimateSize: useCallback(() => DROPDOWN_MENU_ITEM_HEIGHT, []),
    overscan: visibleOptions,
  });

  // init multi-select
  const {
    activeIndex,
    addSelectedItem,
    getDropdownProps,
    getSelectedItemProps,
    removeSelectedItem,
    reset,
    selectedItems,
    setActiveIndex,
    setSelectedItems,
  } = useMultipleSelection<OptionType>({
    // sets focus on the first selected item when the dropdown is initialized
    defaultActiveIndex: 0,
    initialActiveIndex: 0,
    // set the default value(s) when the dropdown is initialized
    ...(props.multiselect && { initialSelectedItems: props.defaultValue ?? [] }),
    // set the selected items when the dropdown is controlled
    ...(props.multiselect && props.value !== undefined && { selectedItems: props.value ?? [] }),
    getA11yRemovalMessage: (props.multiselect && props.getA11yRemovalMessage) ?? (() => ''),
    onSelectedItemsChange: ({ selectedItems: _selectedItems }) =>
      props.multiselect && typeof props.onChange === 'function' && props.onChange(_selectedItems),
    onStateChange: (changes) =>
      onMultiSelectStateChange<OptionType>(changes, activeIndex, selectedItemsContainerRef.current),
    stateReducer: (state, actionAndChanges) => multiSelectReducer<OptionType>(state, actionAndChanges, controlled),
  });

  // init select
  const {
    getItemProps,
    getLabelProps,
    getMenuProps,
    getToggleButtonProps,
    highlightedIndex,
    isOpen,
    selectedItem,
    selectItem,
    reset: resetSelect,
  } = useSelect<OptionType>({
    circularNavigation,
    id,
    items: options,
    // set the default value when the dropdown is initialized
    ...(props.multiselect === false && { initialSelectedItem: props.defaultValue ?? null }),
    // a defined value indicates that the dropdown should be controlled
    // don't set selectedItem if it's not, so that downshift can handle the state
    ...(props.multiselect === false && props.value !== undefined && { selectedItem: props.value }),
    getA11ySelectionMessage,
    getA11yStatusMessage,
    itemToString: (item): string => (item ? item[optionLabelField] ?? '' : ''),
    onSelectedItemChange: ({ selectedItem: _selectedItem }) =>
      props.multiselect === false && typeof props.onChange === 'function' && props.onChange(_selectedItem),
    onStateChange({ type, selectedItem: _selectedItem }) {
      const { ItemClick, MenuBlur, MenuKeyDownEnter, MenuKeyDownSpaceButton } = useSelect.stateChangeTypes;

      if (
        (type === ItemClick || type === MenuBlur || type === MenuKeyDownEnter || type === MenuKeyDownSpaceButton) &&
        props.multiselect &&
        _selectedItem
      ) {
        getIsInSelectedOptions(selectedItems, _selectedItem)
          ? setSelectedItems(selectedItems.filter((item) => !isEqual(item, _selectedItem)))
          : addSelectedItem(_selectedItem);
        selectItem(null);
      }
    },
    stateReducer(state, { type, changes }) {
      const { ItemClick, MenuKeyDownSpaceButton } = useSelect.stateChangeTypes;

      // prevent the menu from being closed when the user selects an item by clicking or pressing space
      if ((type === ItemClick || type === MenuKeyDownSpaceButton) && props.multiselect) {
        return {
          ...changes,
          isOpen: state.isOpen,
          highlightedIndex: state.highlightedIndex,
        };
      }

      return changes;
    },
  });

  const handleWrapperFocus = (e: FocusEvent<HTMLDivElement>) => {
    if (getIsElementFocused(e)) {
      setFocus(true);
      onFocus();
    }
  };

  const handleWrapperBlur = (e: FocusEvent<HTMLDivElement>) => {
    if (getIsElementBlurred(e)) {
      setFocus(false);
      onBlur();
    }
  };

  if (!props.multiselect) {
    // we call the getDropdownProps getter function when multiselect isn't enabled
    // in order to suppress the "You forgot to call the ..." error message thrown by downshift.
    // we only need to apply the getter props to the toggle button when multiselect is enabled.
    getDropdownProps({}, { suppressRefError: true });
  }

  const showClearButtonForSingleSelect = clearable && !props.multiselect && selectedItem;

  // returns the toggle button label based on the dropdown mode
  const getButtonLabel = (): React.ReactNode => {
    let buttonLabel = selectedItem?.[optionLabelField] || placeholder;
    if (props.multiselect) buttonLabel = selectedItems.length > 0 ? null : placeholder;
    return (
      buttonLabel && (
        <span
          className={classNames(
            styles.buttonLabel,
            showClearButtonForSingleSelect && styles.buttonLabelWithClearButton,
          )}
        >
          {buttonLabel}
        </span>
      )
    );
  };

  // screen readers should read the labels in the following order:
  // field label > helper text > error text > toggle button label
  // helper and error texts should only be read if they have been defined
  // prettier-ignore
  const buttonAriaLabel =
    `${getLabelProps().id}${error ? ` ${id}-error` : ''}${helper ? ` ${id}-helper` : ''} ${getToggleButtonProps().id}`;

  // show placeholder if no value is selected
  const showPlaceholder = (props.multiselect && selectedItems.length === 0) || (!props.multiselect && !selectedItem);

  const showIcon =
    (props.icon && props.multiselect === false) || (props.icon && props.multiselect && selectedItems.length === 0);

  return (
    <div
      className={classNames(
        styles.root,
        invalid && styles.invalid,
        disabled && styles.disabled,
        isOpen && styles.open,
        props.multiselect && styles.multiselect,
        customThemeClass,
        className,
      )}
      style={style}
    >
      {/* LABEL */}
      {label && (
        <FieldLabel
          label={label}
          required={required}
          {...getLabelProps()}
          tooltipLabel={tooltipLabel}
          tooltipButtonLabel={tooltipButtonLabel}
          tooltipText={tooltipText}
        />
      )}
      <div className={styles.wrapper} onFocus={handleWrapperFocus} onBlur={handleWrapperBlur}>
        {/* SELECTED ITEMS */}
        {props.multiselect && selectedItems.length > 0 && (
          <SelectedItems<OptionType>
            activeIndex={activeIndex}
            clearable={clearable}
            clearButtonAriaLabel={props.clearButtonAriaLabel}
            dropdownId={id}
            getSelectedItemProps={getSelectedItemProps}
            hideItems={!hasFocus}
            onClear={() => {
              reset();
              toggleButtonRef.current.focus();
            }}
            onRemove={removeSelectedItem}
            optionLabelField={optionLabelField}
            removeButtonAriaLabel={props.selectedItemRemoveButtonAriaLabel}
            selectedItems={selectedItems}
            selectedItemSrLabel={props.selectedItemSrLabel}
            selectedItemsContainerRef={selectedItemsContainerRef}
            setActiveIndex={setActiveIndex}
          />
        )}
        {/* TOGGLE BUTTON */}
        <button
          type="button"
          {...getToggleButtonProps({
            'aria-owns': getMenuProps().id,
            'aria-labelledby': buttonAriaLabel,
            // add downshift dropdown props when multiselect is enabled
            ...(props.multiselect && { ...getDropdownProps({ preventKeyAction: isOpen, ref: toggleButtonRef }) }),
            ...(invalid && { 'aria-invalid': true }),
            disabled,
            className: classNames(styles.button, showPlaceholder && styles.placeholder),
            ...(!props.multiselect && { ref: toggleButtonRef }),
          })}
        >
          {showIcon && (
            <span className={styles.icon} aria-hidden>
              {props.icon}
            </span>
          )}
          {getButtonLabel()}
          <IconAngleDown className={styles.angleIcon} aria-hidden />
        </button>
        {showClearButtonForSingleSelect && (
          <button
            type="button"
            className={styles.clearButton}
            onClick={() => {
              toggleButtonRef.current.focus();
              resetSelect();
            }}
            aria-label={props.clearButtonAriaLabel}
          >
            <IconCrossCircle />
          </button>
        )}
        {/* MENU */}
        <DropdownMenu<OptionType>
          getItemProps={(item, index, selected, optionDisabled, virtualRow) =>
            getItemProps({
              item,
              index,
              disabled: optionDisabled,
              className: classNames(
                styles.menuItem,
                highlightedIndex === index && styles.highlighted,
                selected && styles.selected,
                optionDisabled && styles.disabled,
                virtualized && styles.virtualized,
              ),
              // apply styles for virtualization to menu items
              ...(virtualRow && {
                style: {
                  transform: `translateY(${virtualRow.start}px`,
                },
                ref: virtualRow.measureRef,
              }),
            })
          }
          isOptionDisabled={isOptionDisabled}
          menuProps={getMenuProps({
            ...(props.multiselect && { 'aria-multiselectable': true }),
            ...(required && { 'aria-required': true }),
            style: { maxHeight: DROPDOWN_MENU_ITEM_HEIGHT * visibleOptions },
            ref: menuRef,
          })}
          menuStyles={styles}
          multiselect={props.multiselect}
          open={isOpen}
          optionLabelField={optionLabelField}
          options={options}
          selectedItem={selectedItem}
          selectedItems={selectedItems}
          virtualizer={virtualized && virtualizer}
          visibleOptions={visibleOptions}
        />
      </div>
      {/* INVALID TEXT */}
      {invalid && error && (
        <div id={`${id}-error`} className={styles.errorText} aria-hidden>
          <IconAlertCircle className={styles.invalidIcon} />
          {error}
        </div>
      )}
      {/* HELPER TEXT */}
      {helper && (
        <div id={`${id}-helper`} className={styles.helperText} aria-hidden>
          {helper}
        </div>
      )}
    </div>
  );
};
Select.defaultProps = {
  multiselect: false,
};
