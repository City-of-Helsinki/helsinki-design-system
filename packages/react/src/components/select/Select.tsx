import { uniqueId } from 'lodash';
import React, { useMemo, createRef } from 'react';

import { SelectProps, SelectMetaData, SelectData } from './types';
import { Container } from './components/Container';
import { Label } from './components/Label';
import { changeChandler } from './dataUpdater';
import { propsToGroups, childrenToGroups, getSelectedOptions, getElementIds } from './utils';
import { DataProvider } from '../dataProvider/DataProvider';
import { SelectedOptionsContainer } from './components/selectedOptions/SelectedOptionsContainer';
import { SelectionsAndListsContainer } from './components/SelectionsAndListsContainer';
import { List } from './components/list/List';
import { ListAndInputContainer } from './components/list/ListAndInputContainer';
import { ArrowButton } from './components/selectedOptions/ArrowButton';
import { ButtonWithSelectedOptions } from './components/selectedOptions/ButtonWithSelectedOptions';
import { ClearButton } from './components/selectedOptions/ClearButton';
import { ErrorNotification } from './components/Error';
import { AssistiveText } from './components/AssistiveText';
import { createTextProvider } from './texts';

export function Select({
  options,
  open,
  groups,
  icon,
  required,
  onChange,
  children,
  id,
  disabled,
  texts,
  invalid,
}: SelectProps) {
  const initialData = useMemo<SelectData>(() => {
    return {
      groups: propsToGroups({ options, groups }) || childrenToGroups(children) || [],
      open: !!open,
      required: !!required,
      invalid: !!invalid,
      disabled: !!disabled,
      onChange,
    };
  }, [options, open, groups, onChange, disabled, invalid, required]);

  const metaData = useMemo((): SelectMetaData => {
    const containerId = `${id || uniqueId('hds-select-')}`;
    return {
      lastToggleCommand: 0,
      lastClickedOption: undefined,
      icon,
      refs: {
        listContainer: createRef<HTMLDivElement>(),
        list: createRef<HTMLUListElement>(),
        selectContainer: createRef<HTMLDivElement>(),
        selectionButton: createRef<HTMLButtonElement>(),
      },
      selectedOptions: getSelectedOptions(initialData.groups),
      elementIds: getElementIds(containerId),
      textProvider: createTextProvider(texts),
    };
  }, [id, initialData.groups]);

  return (
    <DataProvider<SelectData, SelectMetaData> initialData={initialData} metaData={metaData} onChange={changeChandler}>
      <Container>
        <Label />
        <SelectionsAndListsContainer>
          <SelectedOptionsContainer>
            <ButtonWithSelectedOptions />
            <ClearButton />
            <ArrowButton />
          </SelectedOptionsContainer>
          <ListAndInputContainer>
            <List />
          </ListAndInputContainer>
        </SelectionsAndListsContainer>
        <ErrorNotification />
        <AssistiveText />
      </Container>
    </DataProvider>
  );
}
