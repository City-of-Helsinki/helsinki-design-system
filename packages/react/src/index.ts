export * from './icons';
export * from './components';
export * from './ssr';
