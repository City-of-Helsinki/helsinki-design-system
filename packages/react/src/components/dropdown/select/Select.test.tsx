import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { Select, SelectProps } from './Select';

const label = 'Select';
const options = [
  { label: 'Finland', value: 'FI' },
  { label: 'Sweden', value: 'SV' },
  { label: 'Norway', value: 'NO' },
  { label: 'Botswana', value: 'BW' },
  { label: 'Bolivia', value: 'BO' },
];
const defaultProps: SelectProps<{ label: string; value: string }> = {
  label,
  options,
  clearButtonAriaLabel: 'Clear all selections',
  selectedItemRemoveButtonAriaLabel: 'Remove item {value}',
  selectedItemSrLabel: 'Selected item {value}',
};

const getWrapper = (props?: unknown) => render(<Select {...defaultProps} {...props} />);

describe('<Select /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = getWrapper();
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders the component without label when aria-labelledby is given', () => {
    const { asFragment } = render(
      <Select
        aria-labelledby="some-external-label-id"
        options={options}
        clearButtonAriaLabel={defaultProps.clearButtonAriaLabel}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = getWrapper();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
