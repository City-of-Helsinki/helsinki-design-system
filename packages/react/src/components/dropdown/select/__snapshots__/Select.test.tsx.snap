// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <label
      class="label "
      for="hds-select-1-toggle-button"
      id="hds-select-1-label"
    >
      Select
    </label>
    <div
      class="wrapper"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="hds-select-1-label hds-select-1-toggle-button"
        aria-owns="hds-select-1-menu"
        class="button placeholder"
        id="hds-select-1-toggle-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="icon s angleIcon"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
            />
            <path
              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
      <ul
        aria-labelledby="hds-select-1-label"
        class="menu"
        id="hds-select-1-menu"
        role="listbox"
        style="max-height: 260px;"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> spec renders the component without label when aria-labelledby is given 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="wrapper"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="hds-select-4-label some-external-label-id hds-select-4-toggle-button"
        aria-owns="hds-select-4-menu"
        class="button placeholder"
        id="hds-select-4-toggle-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="icon s angleIcon"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
            />
            <path
              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
      <ul
        aria-labelledby="hds-select-4-label"
        class="menu"
        id="hds-select-4-menu"
        role="listbox"
        style="max-height: 260px;"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;
